drivers/clock_control/clock_stm32_ll_u3.c

```diff
@@ -0,0 +1,684 @@
+		if (IN_RANGE(STM32_MSIK_RANGE, 0, 3)) {
+			/* Range 0-3 uses RC0 as the clock source */
+			LL_RCC_MSIK_SetClockSource(LL_RCC_MSIK_CLOCK_SOURCE_RC0);
+			switch (STM32_MSIK_RANGE) {
+			case 0:
+				LL_RCC_MSIK_SetClockDivision(LL_RCC_MSIK_CLOCK_SOURCE_RC_DIV_1);
+				break;
+			case 1:
+				LL_RCC_MSIK_SetClockDivision(LL_RCC_MSIK_CLOCK_SOURCE_RC_DIV_2);
+				break;
+			case 2:
+				LL_RCC_MSIK_SetClockDivision(LL_RCC_MSIK_CLOCK_SOURCE_RC_DIV_4);
+				break;
+			case 3:
+				LL_RCC_MSIK_SetClockDivision(LL_RCC_MSIK_CLOCK_SOURCE_RC_DIV_8);
+				break;
+			default:
+				break;
+			}
+		} else {
+			/* Range 4-7 uses RC1 as the clock source */
+			LL_RCC_MSIK_SetClockSource(LL_RCC_MSIK_CLOCK_SOURCE_RC1);
+			switch (STM32_MSIK_RANGE) {
+			case 4:
+				LL_RCC_MSIK_SetClockDivision(LL_RCC_MSIK_CLOCK_SOURCE_RC_DIV_1);
+				break;
+			case 5:
+				LL_RCC_MSIK_SetClockDivision(LL_RCC_MSIK_CLOCK_SOURCE_RC_DIV_2);
+				break;
+			case 6:
+				LL_RCC_MSIK_SetClockDivision(LL_RCC_MSIK_CLOCK_SOURCE_RC_DIV_4);
+				break;
+			case 7:
+				LL_RCC_MSIK_SetClockDivision(LL_RCC_MSIK_CLOCK_SOURCE_RC_DIV_8);
+				break;
+			default:
+				break;
+			}
+
+		}
+		LL_RCC_MSI_SetMSIxClockRange();
+		/*
+		 * For stm32u3 LSE or HSE must be enabled and ready before
+		 * selecting this oscillator as MSIRC0/MSIRC1 input clock.
+		 */
+		/* Enable MSIK */
+		LL_RCC_MSIK_Enable();
+		while (LL_RCC_MSIK_IsReady() == 0U) {
+		}
+		LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_MSIS);
+
+		/* Wait till System clock is ready */
+		while (LL_RCC_GetSysClkSource() != LL_RCC_SYS_CLKSOURCE_MSIS) {
+		}
+
+		LL_RCC_SetAHBPrescaler(LL_RCC_HCLK_SYSCLK_DIV_1);
+		LL_RCC_SetAPB1Prescaler(LL_RCC_APB1_HCLK_DIV_1);
+		LL_RCC_SetAPB2Prescaler(LL_RCC_APB2_HCLK_DIV_1);
+		LL_RCC_SetAPB3Prescaler(LL_RCC_APB3_HCLK_DIV_1);
+
+		__ASSERT(STM32_LSE_ENABLED,
+			"MSIK Hardware auto calibration needs LSE clock activation");
+		if (IN_RANGE(STM32_MSIK_RANGE, 0, 3)) {
+			/*
+			 * LSE or HSE must be enabled and ready before selecting
+			 * this oscillator as MSIRC0 input clock for Range [0 3]
+			 */
+			if (LL_RCC_LSE_IsEnabled() != 0U && LL_RCC_LSE_IsReady() != 0U) {
+				LL_RCC_MSI_RC0_SetPLLInputClk(LL_RCC_MSIPLL0SEL_LSE);
+			} else if (LL_RCC_HSE_IsEnabled() != 0U && LL_RCC_HSE_IsReady() != 0U) {
+				LL_RCC_MSI_RC0_SetPLLInputClk(LL_RCC_MSIPLL0SEL_HSE_OR_HSEDIV2);
+			}
+			LL_RCC_MSI_RC0_PLLmode_Enable();
+			while (LL_RCC_MSI_RC0_PLLmode_IsEnabled() == 0U) {
+			}
+		} else {
+			/*
+			 * LSE or HSE must be enabled and ready before selecting
+			 * this oscillator as MSIRC1 input clock for Range [4 7]
+			 */
+			if (LL_RCC_LSE_IsEnabled() != 0U && LL_RCC_LSE_IsReady() != 0U) {
+				LL_RCC_MSI_RC1_SetPLLInputClk(LL_RCC_MSIPLL1SEL_LSE);
+			} else if (LL_RCC_HSE_IsEnabled() != 0U && LL_RCC_HSE_IsReady() != 0U) {
+				LL_RCC_MSI_RC1_SetPLLInputClk(LL_RCC_MSIPLL1SEL_HSE_OR_HSEDIV2);
+			}
+			LL_RCC_MSI_RC1_PLLmode_Enable();
+			while (LL_RCC_MSI_RC1_PLLmode_IsEnabled() == 0U) {
+			}
+		}
```

Comment:
This whole code block is duplicated with MSIS and can be replaced by a function which would take range as argument (the only variable part)